packages/orm/src/EntityManager.ts

```diff
@@ -1065,52 +1083,62 @@ export class EntityManager<C = unknown> {
         await currentFlushSecret.run({ flushSecret: this.flushSecret }, async () => {
-          let todos = createTodos(pendingEntities);
-
           // Run our hooks
+          let todos = createTodos(pendingEntities);
           await beforeCreate(this.ctx, todos);
           await beforeUpdate(this.ctx, todos);
           await beforeFlush(this.ctx, todos);
-          // Recalc todos in case one of ^ hooks did an em.delete
+          // Recalc todos in case one of ^ hooks did an em.delete that moved an
+          // entity in this pending loop from created/updated to deleted.
           todos = createTodos(pendingEntities);
           await beforeDelete(this.ctx, todos);
 
-          // We defer doing this cascade logic until flush() so that delete() can remain synchronous.
-          await cleanupDeletedRelations(todos);
-
-          recalcDerivedFields(todos);
-
-          // After hooks have run, and potentially updated fields, see if
-          // any async derived values depend on changed fields.
-          //
-          // Note that we defer validation until our run-hooks loop is complete,
-          // so that if a hook creates an entity, we don't invoke reactive validation
-          // rules at this point, until we give the entity a chance to run its own hooks
-          // as part of the next loop.
-          await addReactiveAsyncDerivedValues(todos);
-          // We include this inside the loop, b/c if we change values on a reacted,
-          // not-yet-mutated entity, we'll want to loop around and run its hooks
-          await recalcAsyncDerivedFields(this, todos);
-
-          entitiesToFlush.push(...pendingEntities);
-          pendingEntities = this.entities.filter((e) => e.isPendingFlush && !entitiesToFlush.includes(e));
+          // Call `setField` just to get the column marked as dirty if needed.
+          // This can come after the hooks, b/c if the hooks read any of these
+          // fields, they'd be via the synchronous getter and would not be stale.
+          recalcSynchronousDerivedFields(todos);
+
+          // The hooks could have deleted this-loop or prior-loop entities, so re-cascade again.
+          await this.cascadeDeletes();
+          // The hooks could have changed fields, so recalc again.
+          await this.__data.rm.recalcPendingDerivedValues();
+
+          for (const e of pendingEntities) hooksInvoked.add(e);
+          pendingEntities = this.entities.filter((e) => e.isPendingFlush && !hooksInvoked.has(e));
           this.flushSecret += 1;
         });
       }
 
+      // Recreate todos now that we've run hooks and recalculated fields so know
+      // the full set of entities that will be INSERT/UPDATE/DELETE-d in the database.
+      const entitiesToFlush = [...hooksInvoked];
       const entityTodos = createTodos(entitiesToFlush);
 
       if (!skipValidation) {
         try {
           this._isValidating = true;
-          await addReactiveValidations(entityTodos);
```

Comment:
Since `Todo.asyncFields` went away + `validates` had no reason to be part of the `Todo` data structure, I inlined `addReactiveValidations` into `validateReactiveRules`.